### ISSUE TITLE
UserWarning: Discrepancy between trainable weights and collected trainable weights, did you set `model.trainable` without calling `model.compile` after ?   'Discrepancy between trainable weights and collected trainable'

### ISSUE THREAD

**Issue by Palak-15** [tensorflow/tensorflow#33018]
Please Check my process because I am trying to solve this issue from very long time. I tried every thing on google but it didn't help.

1) build_generator() and build_discriminator() are 2 functions
2)inputs = Input(shape=(MAX_SEQUENCE_LENGTH,lattice))
dis_model=build_discriminator(inputs)
dis_model.summary()
dis_model.compile(optimizer=Adam(0.0001, beta_1=0.5, beta_2=0.9))
n_disc_trainable = len(dis_model.trainable_weights)
print("Discriminator",n_disc_trainable)

3) gen_model=build_generator()
gen_model.trainable = False
gen_model.compile(optimizer=Adam(0.0001, beta_1=0.5, beta_2=0.9),)
n_gen_trainable = len(gen_model.trainable_weights)

4)update_data = np_utils.to_categorical(data[0:10],lattice)
real_samples = Input(shape=update_data.shape[1:])
generator_input_for_discriminator = Input(shape=(seed,))
generated_samples_for_discriminator = gen_model(generator_input_for_discriminator)
discriminator_output_from_generator = dis_model(generated_samples_for_discriminator)
discriminator_output_from_real_samples = dis_model(real_samples)

averaged_samples = RandomWeightedAverage()([real_samples,generated_samples_for_discriminator])
averaged_samples_out = dis_model(averaged_samples)

partial_gp_loss = partial(gradient_penalty_loss,
                          averaged_samples=averaged_samples,
                          gradient_penalty_weight=GRADIENT_PENALTY_WEIGHT)
# Functions need names or Keras will throw an error
partial_gp_loss.__name__ = 'gradient_penalty'

discriminator_model = Model(inputs=[real_samples,
                                    generator_input_for_discriminator],
                            outputs=[discriminator_output_from_real_samples,
                                     discriminator_output_from_generator,
averaged_samples_out])

discriminator_model.compile(optimizer=Adam(0.0001, beta_1=0.5, beta_2=0.9),
                            loss=[Wasserstein_loss,
                                  Wasserstein_loss,
partial_gp_loss],metrics=['accuracy'])
discriminator_model.summary()

5)# PREPARATION FOR GENERATOR MODEL
dis_model.trainable = False
dis_model.compile(optimizer=Adam(0.0001, beta_1=0.5, beta_2=0.9),loss=Wasserstein_loss,metrics=['accuracy'])

gen_model.trainable = True
gen_model.compile(optimizer=Adam(0.0001, beta_1=0.5, beta_2=0.9),loss=Wasserstein_loss,metrics=['accuracy'])

# COMPILE AND INITIALIZE GENERATOR MODEL
generator_input = Input(shape=(seed,))
generator_layers = gen_model(generator_input)
discriminator_layers_for_generator = dis_model(generator_layers)
generator_model = Model(inputs=[generator_input],
                        outputs=[discriminator_layers_for_generator])

generator_model.compile(optimizer=Adam(0.0001, beta_1=0.5, beta_2=0.9),
loss=Wasserstein_loss,metrics=['accuracy'])
generator_model.summary()

print("BUILD DISCRIMINATOR MODEL TRAINABLE STATUS",dis_model.trainable)
dis_model.trainable = True
dis_model.compile(optimizer=Adam(0.0001, beta_1=0.5, beta_2=0.9),loss=Wasserstein_loss,metrics=['accuracy'])

positive_y = np.ones((batch_size, 1), dtype=np.float32)
negative_y = -positive_y
dummy_y = np.zeros((batch_size, 1), dtype=np.float32)

6) Then a training loop  ( Train on batch - both discriminator_model and generator_model)

**Comment by gadagashwini** [comment]
@Palak-15, Please provide the information asked in the [Template](https://github.com/tensorflow/tensorflow/issues/new/choose). Thanks!

**Comment by Palak-15** [comment]
@gadagashwini,
Please find the detail and this is not hardware issue for your kind information. There is no need of that template.
**System information**
- I already provided you my code.
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow version (use command below): Tried all version 
**Describe the current behavior**- Problem in wgan, weights are different. 

**Describe the expected behavior**- They should be same.

**Code to reproduce the issue**- I have already given
**Other info / logs**- Already given

**Comment by gadagashwini** [comment]
@Palak-15, Please provide the intended code snippet to reproduce the issue. Thanks!

**Comment by Palak-15** [comment]
@gadagashwini ,

Excuse me. I already provided it, main code. are you new on github or on tensorflow?

**Comment by gadagashwini** [comment]
@Palak-15, I saw the above code and i tried replicating the reported issue but looks like the code is not properly intended. Please provide the complete standalone code to reproduce the issue. Thanks!
